bin/remigrate_visit_logs.ts

```diff
@@ -0,0 +1,186 @@
+            'visit_activity.visit_activity_id',
+            'visit_log.visit_activity_id')
+          .where('user_account.qr_code', visitor.qrCode)
+          .whereBetween('visit_log.created_at', [new Date('2018-12-13T19:00:00'), new Date()]);
+
+        // 2.2 Get list of logs from source DB (_before data sync), B
+        const sourceLogsPreSync = await source
+          .select({
+            qrCode: 'users.hash',
+            activity: 'activities.name',
+            createdAt: 'visits.date',
+          })
+          .from('visits')
+          .innerJoin(
+            'users',
+            'users.id',
+            'visits.usersid')
+          .leftOuterJoin(
+            'activities',
+            'activities.id',
+            'visits.activitiesid')
+          .where('users.hash', visitor.qrCode)
+          .whereBetween('visits.date', [new Date(0), new Date('2018-12-13T19:00:00')]);
+
+        const logs = sourceLogsPreSync.concat(trgtLogsPostSync);
+
+        const pl = await Promise.all(logs.map(async (log: any) => {
+          // Need to pre-fetch activities instead of using sub-queries
+          // in order to prevent "multiple rows" or "null" errors
+          const activities = await trx('visit_activity')
+            .select(['visit_activity_id', 'deleted_at'])
+            .where({
+              visit_activity_name: log.activity,
+              organisation_id: org.id,
+            });
+
+          // If multiple activities are returned, this is probably
+          // because one has been deleted (and then re-added).
+          // Select the non-deleted one.
+          const activityId = ((activities.length > 1
+            ? activities.find((a: any) => a.deleted_at !== null)
+            : activities[0]) || {}).visit_activity_id;
+
+          // Ignoring junk visits
+          // // This is a real user who registered clearly junk visits,
+          // // since the visit activities came from a test organisation
+          // // (see https://github.com/TwinePlatform/data-sync/issues/39#issuecomment-451682133)
+          if (!activityId && visitor.qrCode === ignoreNullAct) {
+            return null;
+          }
+
+          return {
+            user_account_id: log.userId || trx('user_account')
+              .select('user_account_id')
+              .where('qr_code', log.qrCode),
+            visit_activity_id: activityId,
+            created_at: log.createdAt,
+          };
+        }));
+
+        // needed because of line 138
+        const payload = pl.filter((x) => x !== null);
+
+        console.log(
+          `${visitor.id}@${org.name} :: `,
+          `Source logs: ${sourceLogsPreSync.length} | `,
+          `Target logs: ${trgtLogsPostSync.length}`,
+          `Inserting ${payload.length} logs`
+        );
+
+        try {
+          console.log(`Visitor ${visitor.qrCode} has ${trgtLogsPostSync.length} new logs`);
+          await trx('visit_log').del().where('user_account_id', visitor.id);
```

Comment:
What's going on here?

Reply:
Deleting all existing visit logs of the given visitor (re-added, along with any missing logs, on the next line).